lib/semantics/semantics.cc

```diff
@@ -228,6 +230,35 @@ void DoDumpSymbols(std::ostream &os, const Scope &scope, int indent) {
   --indent;
 }
 
+void Semantics::DumpSymbolsSources(std::ostream &os) {
+  NameToSymbolMap symbols;
+  GetAllSymbols(context_.globalScope(), symbols);
+  for (const auto pair : symbols) {
+    auto sourceInfo{cooked_.GetSourcePositionRange(pair.second->name())};
+    if (sourceInfo) {
+      os << pair.second->name().ToString() << ": " << sourceInfo->first.line
+         << ", " << sourceInfo->first.column << "-" << sourceInfo->second.column
+         << "\n";
+    } else if (pair.second->has<semantics::UseDetails>()) {
+      os << pair.second->name().ToString() << ": "
+         << pair.second->GetUltimate().owner().symbol()->name().ToString()
+         << "\n";
+    }
+  }
+}
+
+void GetAllSymbols(const Scope &scope, NameToSymbolMap &symbols) {
```

Comment:
Might be better as a `const` member function.

Reply:
I think this function is too specific to be a member function of Semantics, because its purpose is to find all symbol names and not collect duplicates. I will rename it to be more specific.